File: Shortest and Lexicographically Smallest Beautiful String/Shortest and Lexicographically Smallest Beautiful String.java
// Solution for Shortest and Lexicographically Smallest Beautiful String in JAVA

class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        int left = 0;
        int ones = 0;
        String best = "";

        for (int right = 0; right < s.length(); right++) {

            if (s.charAt(right) == '1') {
                ones++;
            }

            // More than k ones -> move left
            while (ones > k) {
                if (s.charAt(left) == '1') {
                    ones--;
                }
                left++;
            }

            // Remove leading zeros
            while (left <= right && s.charAt(left) == '0') {
                left++;
            }

            // Exactly k ones
            if (ones == k) {
                String candidate = s.substring(left, right + 1);

                if (best.isEmpty()
                        || candidate.length() < best.length()
                        || (candidate.length() == best.length()
                            && candidate.compareTo(best) < 0)) {
                    best = candidate;
                }
            }
        }

        return best;
    }
}

File: Shortest and Lexicographically Smallest Beautiful String/Shortest and Lexicographically Smallest Beautiful String.py
# Solution for Shortest and Lexicographically Smallest Beautiful String in PY

class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        left = 0
        ones_count = 0
        best = ""

        for right in range(len(s)):
            if s[right] == '1':
                ones_count += 1

            # Remove extra 1s
            while ones_count > k:
                if s[left] == '1':
                    ones_count -= 1
                left += 1

            # Remove leading zeros
            while left <= right and s[left] == '0':
                left += 1

            # Exactly k ones
            if ones_count == k:
                candidate = s[left:right + 1]

                if (
                    not best
                    or len(candidate) < len(best)
                    or (len(candidate) == len(best) and candidate < best)
                ):
                    best = candidate

        return best
